torch_xla/csrc/tensor_util.cpp

```diff
@@ -947,8 +949,13 @@ std::vector<torch::lazy::BackendDataPtr> CreateTensorsData(
       // Shards the input tensors with padding, to split evenly.
       // The execution requires consistent shard sizes, and the zero-padded
       // values should be ignored.
-      std::vector<at::Tensor> local_shards = ShardingUtil::ShardTensor(
-          tensors[i], sharding, local_devices, /*padded=*/true);
+      std::vector<at::Tensor> local_shards =
+          ShardingUtil::ShardTensor(tensors[i], shardings[i], local_devices,
+                                    /*padded=*/true);
+      if (minibatch) {  // change global shape as tensor is already sharded
+                        // accross batch dimesion.
+        shape = shardings[i]->shape;
+      }
       new_handles.push_back(ShardingUtil::CreateShardedData(
           local_shards, local_devices, shape, sharding));
```

Comment:
This is turning into a larger refactor, but can we also just pass the `ShardingSpecPtr` into `CreateShardedData`?

Then we won't need L955-958 and L939-948